contracts/src/registry/PermissionedRegistry.sol

```diff
@@ -2,55 +2,295 @@
+
+    uint256 private constant ROLE_RENEW = 1 << 1;
+    uint256 private constant ROLE_RENEW_ADMIN = ROLE_RENEW << 128;
+
+    uint256 private constant ROLE_SET_SUBREGISTRY = 1 << 2;
+    uint256 private constant ROLE_SET_SUBREGISTRY_ADMIN = ROLE_SET_SUBREGISTRY << 128;
+
+    uint256 private constant ROLE_SET_RESOLVER = 1 << 3;
+    uint256 private constant ROLE_SET_RESOLVER_ADMIN = ROLE_SET_RESOLVER << 128;
+
+    uint256 private constant ROLE_SET_TOKEN_OBSERVER = 1 << 4;
+    uint256 private constant ROLE_SET_TOKEN_OBSERVER_ADMIN = ROLE_SET_TOKEN_OBSERVER << 128;
 
-abstract contract PermissionedRegistry is BaseRegistry {
-    uint96 public constant FLAGS_MASK = 0x7;
-    uint96 public constant FLAG_SUBREGISTRY_LOCKED = 0x1;
-    uint96 public constant FLAG_RESOLVER_LOCKED = 0x2;
-    uint96 public constant FLAG_FLAGS_LOCKED = 0x4;
+    modifier onlyNonExpiredTokenRoles(uint256 tokenId, uint256 roleBitmap) {
+        _checkRoles(getTokenIdResource(tokenId), roleBitmap, _msgSender());
+        (, uint64 expires, ) = datastore.getSubregistry(tokenId);
+        if (expires < block.timestamp) {
+            revert NameExpired(tokenId);
+        }
+        _;
+    }
+
+    constructor(IRegistryDatastore _datastore, IRegistryMetadata _metadata, uint256 _deployerRoles) BaseRegistry(_datastore) MetadataMixin(_metadata) {
+        _grantRoles(ROOT_RESOURCE, _deployerRoles, _msgSender(), false);
+
+        if (address(_metadata) == address(0)) {
+            _updateMetadataProvider(new SimpleRegistryMetadata());
+        }
+    }
 
-    constructor(IRegistryDatastore _datastore) BaseRegistry(_datastore) {
+    function uri(uint256 tokenId ) public view override returns (string memory) {
+        return tokenURI(tokenId);
     }
 
-    function _setFlags(uint256 tokenId, uint96 _flags)
-        internal
-        withSubregistryFlags(tokenId, FLAG_FLAGS_LOCKED, 0)
-        returns(uint96 newFlags)
+    function ownerOf(uint256 tokenId)
+        public
+        view
+        virtual
+        override(ERC1155Singleton, IERC1155Singleton)
+        returns (address)
     {
-        (address subregistry, uint96 oldFlags) = datastore.getSubregistry(tokenId);
-        newFlags = oldFlags | (_flags & FLAGS_MASK);
-        if (newFlags != oldFlags) {
-            datastore.setSubregistry(tokenId, subregistry, newFlags);
+        (, uint64 expires, ) = datastore.getSubregistry(tokenId);
+        if (expires < block.timestamp) {
+            return address(0);
         }
+        return super.ownerOf(tokenId);
+    }
+
+    function register(string calldata label, address owner, IRegistry registry, address resolver, uint256 roleBitmap, uint64 expires)
+        public
+        override
+        onlyRootRoles(ROLE_REGISTRAR)
+        returns (uint256 tokenId)
+    {
+        uint64 oldExpiry;
+        uint32 tokenIdVersion;
+        (tokenId, oldExpiry, tokenIdVersion) = this.getNameData(label);
```

Comment:
Calling via `this` makes it an external call, which is less efficient. Is this needed here?